third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp

```diff
@@ -43,45 +61,85 @@ class TritonAMDGPUReorderInstructionsPass
   void runOnOperation() override {
     ModuleOp m = getOperation();
     mlir::DominanceInfo dom(m);
-    // Sink conversions into loops when they will increase
-    // register pressure
+
+    // Sink shared memory loads and layout conversions into loops to decrease
+    // register pressure when possible.
     DenseMap<Operation *, Operation *> opToMove;
-    auto moveAfter = [](Operation *lhs, Operation *rhs) {
-      lhs->moveAfter(rhs);
-    };
     m.walk([&](Operation *op) {
-      if (!willIncreaseRegisterPressure(op))
+      if (!isLocalLoadOrDotLayoutConversion(op))
         return;
-      auto user_begin = op->user_begin();
-      auto user_end = op->user_end();
-      if (std::distance(user_begin, user_end) != 1)
+      if (!op->hasOneUse())
         return;
-      if (user_begin->getParentOfType<scf::ForOp>() ==
+      Operation *user = *op->getUsers().begin();
+      if (user->getParentOfType<scf::ForOp>() ==
           op->getParentOfType<scf::ForOp>())
         return;
-      opToMove.insert({op, *user_begin});
+      opToMove.insert({op, user});
     });
     for (auto &kv : opToMove)
       kv.first->moveBefore(kv.second);
+    opToMove.clear();
+
     // Move LocalLoadOp and LocalAllocOp immediately after their operands.
+    // This enables issuing them as early as possible.
     m.walk([&](Operation *op) {
-      if (!isa<triton::gpu::LocalLoadOp, triton::gpu::LocalAllocOp>(op)) {
-        return;
-      }
-      Operation *argOp = op->getOperand(0).getDefiningOp();
-      if (!argOp)
+      if (!isa<ttg::LocalLoadOp, ttg::LocalAllocOp>(op) ||
+          op->getNumOperands() < 1)
         return;
-      moveAfter(op, argOp);
+      if (Operation *argOp = op->getOperand(0).getDefiningOp())
+        op->moveAfter(argOp);
     });
-    // Move transpositions just after their definition
-    opToMove.clear();
+
+    // Move transpositions just after their definition.
     m.walk([&](triton::TransOp op) {
-      Operation *argOp = op.getSrc().getDefiningOp();
-      if (!argOp)
-        return;
-      moveAfter(op, argOp);
+      if (Operation *argOp = op.getSrc().getDefiningOp())
+        op->moveAfter(argOp);
     });
-    return;
+
+    SmallVector<Operation *> moveOps;
+    // Move global loads early to prefetch. This may increase register pressure
+    // but it enables issuing global loads early.
+    m.walk([&](triton::LoadOp op) { moveOps.push_back(op); });
+    // Move local_stores early if dependence distance greater than
+    // one iteration. Best perf on GEMM when these precede global loads.
+    m.walk([&](ttg::LocalStoreOp op) { moveOps.push_back(op); });
+
+    for (auto op : moveOps) {
+      // Gather use-def chain in block.
+      Block *block = op->getBlock();
+      bool leadsToLoad = false;
```

Comment:
Can you add some comments to further explain what `leadsToLoad` is for? It's not straightforward here.